dropwizard-index-page/src/main/java/com/palantir/indexpage/FileSystemIndexPage.java

```diff
@@ -15,63 +14,90 @@
 import com.google.common.io.Files;
 import java.io.File;
 import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.Reader;
 import java.io.StringWriter;
 import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Used to represent the index page loaded from the file system.
  */
 public final class FileSystemIndexPage implements IndexPage {
+    private static final Logger log = LoggerFactory.getLogger(FileSystemIndexPage.class);
 
     private final Map<String, String> context;
     private final DefaultMustacheFactory factory;
     private final File file;
 
-    private Optional<String> content;
-    private long lastModified;
+    private Optional<IndexPageState> state = Optional.absent();
 
     public FileSystemIndexPage(Map<String, String> context, File file) {
-        checkNotNull(context);
-        checkArgument(file != null && file.exists());
+        checkNotNull(context, "context");
+        checkNotNull(file, "file");
 
         this.context = ImmutableMap.copyOf(context);
         this.factory = new DefaultMustacheFactory();
         this.file = file;
 
         try {
-            loadFile();
-        } catch (FileNotFoundException e) {
-            throw new RuntimeException("Index page cannot be found at " + file.toString(), e);
+            refresh();
+        } catch (Exception e) {
+            log.error("Failed to preload index page {}", file, e);
         }
     }
 
     @Override
     public Optional<String> getContent() {
-        try {
-            checkAndReloadFile();
-        } catch (FileNotFoundException e) {
+        refresh();
+        return state.transform(input -> input.content);
+    }
 
-            // the file could be deleted during run time, so return absent in order to handle the behavior gracefully
-            return Optional.absent();
+    private synchronized void refresh() {
+        if (!state.isPresent() || state.get().lastModified != file.lastModified()) {
+            log.debug("Reloading index page {}", file.lastModified());
+            state = loadFile();
         }
+    }
 
-        return content;
+    private Optional<IndexPageState> loadFile() {
+        // Guarantee that the modification time did not change between start and end of compilation
+        for (;;) {
+            long preLastModified = file.lastModified();
+            Optional<String> optContent = compileContent();
```

Comment:
The point of getting `lastModified` twice is precisely so this is outside the if-check; I'm bookending the compilation step by getting modification time before and after, and checking equality. This provides stronger checks against the case where compilation is very slow (or where you happen to hit save mid-compile), to avoid persistent-stale results.

More generally, think of it as an attempt to make sure that the returned (compiled content, timestamp) pair is actually self-consistent.